Add documentation for snapshot mode and dual session (#207)
ros2/ros2_tracing: README.md

Reviewer: Can you modify the content you added to have _exactly_ 1 sentence per line to fit with the rest of the document?

> For more information, refer to the [design document](./doc/design_ros_2.md#general-guidelines).
> The [launch file action](#launch-file-trace-action) is designed to automatically start tracing before the application launches.

> [!TIP]

> > Configuring a tracing session in [snapshot mode](#tracing-in-snapshot-mode) or [dual session mode](#dual-session-tracing) stores trace data in the memory without writing to disk until demanded.

Reviewer: Removing the "the" before "memory" here makes it sound better to me, but I'll let you decide.

```suggestion
> Configuring a tracing session in [snapshot mode](#tracing-in-snapshot-mode) or [dual session mode](#dual-session-tracing) stores trace data in memory without writing to disk until demanded.
```

Author: I agree, making this change!

> > This can be leveraged to start tracing after application launch without losing initialization data, eliminating the need to start tracing before application launch and preventing accumulation of unwanted trace files when not actively analyzing.

Reviewer: What do you think of this?

```suggestion
> This can be leveraged to start tracing after application launch without losing initialization data, eliminating the need to actively record trace data before application launch and preventing accumulation of unwanted trace files when not actively analyzing.
```

Author: I think this is good.

I will also change this:
> Configuring a tracing session in [snapshot mode](#tracing-in-snapshot-mode) or [dual session mode](#dual-session-tracing) **records** trace data in the memory without writing to disk until demanded.

to:
> Configuring a tracing session in [snapshot mode](#tracing-in-snapshot-mode) or [dual session mode](#dual-session-tracing) **stores** trace data in the memory without writing to disk until demanded.

to avoid confusion regarding the term "recording"

Reviewer: great idea!

Author: should we stick to this convention throughout the documentation?

start tracing --> starting a tracing session, whether or not the data is being written to the disk
store trace data in memory (or sub-buffers) --> primarily for snapshot sessions
record trace data --> writing trace data to the disk, when we take a snapshot for snapshot sessions or when a normal tracing session is active

Reviewer: I would only use this kind of more precise language when talking about dual session or snapshot tracing. For normal tracing sessions/use-cases (the rest of this document), I would leave it as-is to keep it simple. Does that make sense?

Author: yeah!

> > This can be leveraged to start recording traces after application launch without losing initialization data, eliminating the need to record trace data before application launch and preventing accumulation of unwanted trace files when not actively analyzing.

Reviewer: I know what you mean, but the sentence is a bit repetitive. It reads like: "this can be leveraged to _avoid doing X_, _eliminating the need to do X_ and [...]."

I think this would help make it simpler and make the point/intention clearer:

```suggestion
> This can be leveraged to start recording traces as needed after application launch without losing initialization data, eliminating the need to start recording all trace data before application launch and preventing accumulation of unwanted trace files when not actively analyzing.
```

You must [install the kernel tracer](#building) if you want to enable [kernel](https://lttng.org/docs/v2.13/#doc-tracing-the-linux-kernel) events (`events_kernel` in Python, `events-kernel` in XML or YAML) or syscalls (`syscalls` in Python, XML, or YAML).
If you have installed the kernel tracer, use kernel tracing, and still encounter an error here, make sure to [add your user to the `tracing` group](#tracing).

## Tracing in snapshot mode

> By default, tracing sessions write trace data continuously to disk. Tracing sessions in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode) store trace data in memory and only write to disk when a [snapshot is taken](https://lttng.org/docs/v2.13/#doc-taking-a-snapshot). When memory buffers fill up, the oldest data is discarded, maintaining a rolling history whose duration can be controlled by configuring sub-buffer size. This is useful for capturing trace data only when something interesting occurs, avoiding continuous disk writes.

Reviewer: I would mention that a snapshot session has even less of a runtime performance impact here:

```suggestion
By default, tracing sessions write trace data continuously to disk. Tracing sessions in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode) store trace data in memory and only write to disk when a [snapshot is taken](https://lttng.org/docs/v2.13/#doc-taking-a-snapshot). When memory buffers fill up, the oldest data is discarded, maintaining a rolling history whose duration can be controlled by configuring sub-buffer size. This is useful for capturing trace data only when something interesting occurs, avoiding continuous disk writes and thus lowering the runtime performance impact even more.
```

> By default, tracing sessions write trace data continuously to disk. Tracing sessions in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode) store trace data in memory and only write to disk when a [snapshot is taken](https://lttng.org/docs/v2.13/#doc-taking-a-snapshot). When memory buffers fill up, the oldest data is discarded, maintaining a rolling history whose duration can be controlled by configuring sub-buffer size. This is useful for capturing trace data only when something interesting occurs, avoiding continuous disk writes.

Reviewer: > When memory buffers fill up, the oldest data is discarded

That's what the LTTng docs say: https://lttng.org/docs/v2.13/#doc-taking-a-snapshot

> 1. Create a recording session in [snapshot mode](https://lttng.org/docs/v2.13/#doc-snapshot-mode):
>     ```
>     $ lttng create my-session --snapshot
>     ```
>     The [event record loss mode](https://lttng.org/docs/v2.13/#doc-channel-overwrite-mode-vs-discard-mode) of [channels](https://lttng.org/docs/v2.13/#doc-channel) created in this mode is automatically set to [_overwrite_](https://lttng.org/docs/v2.13/#doc-overwrite-mode).

However, I don't think it's the case for us because we manually create channels using the lttng-ctl API (through lttngpy) and unconditionally use the discard (new events) mode instead of the overwrite (old events) mode:

- https://github.com/ros2/ros2_tracing/blob/05cd9bad40617f46e72f068e1313dfc72f7860ac/tracetools_trace/tracetools_trace/tools/lttng_impl.py#L346
- https://github.com/ros2/ros2_tracing/blob/05cd9bad40617f46e72f068e1313dfc72f7860ac/tracetools_trace/tracetools_trace/tools/lttng_impl.py#L381

I would keep the documentation as-is, because it's what we should do, but could you open an issue to use discard mode for channels of a snapshot session?

Author: Ah! I missed this earlier.

Author: I think this is a minor change, should I open a PR directly instead?

Reviewer: you can if you want, or you can leave it for later (issue and/or PR), after the final submission and your exams

> By default, tracing sessions write trace data continuously to disk. Tracing sessions in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode) store trace data in memory and only write to disk when a [snapshot is taken](https://lttng.org/docs/v2.13/#doc-taking-a-snapshot). When memory buffers fill up, the oldest data is discarded, maintaining a rolling history whose duration can be controlled by configuring sub-buffer size. This is useful for capturing trace data only when something interesting occurs, avoiding continuous disk writes.

Reviewer: > maintaining a rolling history whose duration can be controlled by configuring sub-buffer size

I would say "size" instead of "duration," since we directly control the _size_ of the data (through the sub-buffer size) as opposed to directly controlling the _duration_ of the data we collect, even if they're kind of linked.

### Trace command

Use the `--snapshot-mode` option while starting a tracing session to configure it in snapshot mode:

```

> $ ros2 trace --snapshot-mode  # requires user interaction

Reviewer: ```suggestion
$ ros2 trace --snapshot-mode   # requires user interaction
```

Reviewer: my bad, I wanted it to use 2 spaces, but it's already 2 spaces. Ignore this

```

```
$ ros2 trace start session_name --snapshot-mode
```

By default all ROS 2 tracepoints are enabled. Run the command with `-h` for more information.

Use `record_snapshot` to take a snapshot and write the trace data to disk:

```
$ ros2 trace record_snapshot session_name
```

All other session control commands work as usual:

```
$ ros2 trace pause session_name
$ ros2 trace resume session_name
$ ros2 trace stop session_name
```

### Launch file trace action

> Set `snapshot_mode = True` in the `Trace` action to configure the tracing session in snapshot mode using launch files. The session starts when launching the launch file and ends when it exits or when terminated.

Reviewer: Since `kwargs` are usually set as `name=value` without spaces:

```suggestion
Set `snapshot_mode=True` in the `Trace` action to configure the tracing session in snapshot mode using launch files. The session starts when launching the launch file and ends when it exits or when terminated.
```

```
$ ros2 launch tracetools_launch example_snapshot_mode.launch.py
```

Use `record_snapshot` to take a snapshot and write the trace data to disk, as stated earlier.

> [!NOTE]
> In snapshot mode, high-frequency runtime events may overwrite initialization trace data in memory before a snapshot is taken, potentially making the trace data unusable.

> >  This is more likely to occur if sub-buffer sizes are not configured appropriately.

Reviewer: ```suggestion
> This is more likely to occur if sub-buffer sizes are not configured appropriately.
```

> >  This is more likely to occur if sub-buffer sizes are not configured appropriately.

Reviewer: Instead of "not configured appropriately," we can say "too small," right? That would be more straightforward for users, otherwise they may ask, "what's an appropriate configuration, then?"

Author: yes, I think that is a better way to convey this issue

> > Support for configuring separate channels for initialization and runtime tracepoints is planned to address this limitation (see[#199](https://github.com/ros2/ros2_tracing/issues/199)).

Reviewer: ```suggestion
> Support for configuring separate channels for initialization and runtime tracepoints is planned to address this limitation (see [#199](https://github.com/ros2/ros2_tracing/issues/199)).
```

## Dual session tracing

> Dual session mode solves the problem of losing initialization trace data by using two separate tracing sessions: one for initialization events in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode), and another normal tracing session for runtime events. This allows tracing to be started at any point without losing initialization data.

Reviewer: What do you think?

```suggestion
Dual session mode solves the problem of losing initialization trace data by using two separate tracing sessions: one for initialization events in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode), and another normal tracing session for runtime events. This allows actively starting to trace and record at any point without losing initialization data.
```

Author: Related to my coment: https://github.com/ros2/ros2_tracing/pull/207#discussion_r2334980607

How about "This allows starting to actively record trace data at any point without losing initialization data"?

Reviewer: Yeah, works for me! That wording is better

> Use the Trace action to start the initialization session in snapshot mode:

Reviewer: ```suggestion
Use the `Trace` action to start the initialization session in snapshot mode:
```

```
$ ros2 launch tracetools_launch example_dual_session.launch.py
```

> Then use the trace commands with `--dual_session` option to start and control the sessions.

Reviewer: `--dual_session` :arrow_right: `--dual-session` here and for 5 other instances below

> To take a snapshot of the initialization session and start the runtime session:

Reviewer: ```suggestion
To take a snapshot of the initialization session and start the runtime session:

```

```

> $ ros2 trace -s session_name --dual_session # requires user interaction

Reviewer: ```suggestion
$ ros2 trace -s session_name --dual_session  # requires user interaction
```

Author: The convention is to use 2 spaces or 3? in this comment (https://github.com/ros2/ros2_tracing/pull/207#discussion_r2334229742) you suggested using 3 spaces

Reviewer: Oops, it's 2 spaces, sorry! Ignore the other one

```

```
$ ros2 trace start session_name --dual_session
```

> By default tracepoints corresponding to initialization ROS 2 events are enabled in initialization session and all ROS 2 tracepoints are enabled in runtime session. The snapshot trace will be written to `~/.ros/tracing/session_name/snapshot` and runtime trace will be written to `~/.ros/tracing/session_name/runtime`. Run the commands with `-h` for more information.

Reviewer: ```suggestion
By default tracepoints corresponding to initialization ROS 2 events are enabled for the initialization (snapshot) session and all ROS 2 tracepoints are enabled for the runtime session. The snapshot trace will be written to `~/.ros/tracing/session_name/snapshot` and runtime trace will be written to `~/.ros/tracing/session_name/runtime`. Run the commands with `-h` for more information.
```

> Other session control commands work as follows:

Reviewer: ```suggestion
Other session control commands work as follows:

```

```
$ ros2 trace pause session_name --dual_session   # Pause runtime session after starting
$ ros2 trace resume session_name --dual_session  # Take a snapshot of initialization session
                                                   and resume runtime session after pausing
$ ros2 trace stop session_name --dual_session    # Stop runtime session
```

> 

Reviewer: ```suggestion
```

> [!NOTE]

> > The session name used in both the Trace action and the trace commands must be the same.

Reviewer: ```suggestion
> In dual session mode, the session name used in both the `Trace` action and the trace commands must be the same.
```

You must [install the kernel tracer](#building) if you want to enable [kernel](https://lttng.org/docs/v2.13/#doc-tracing-the-linux-kernel) events (`events_kernel` in Python, `events-kernel` in XML or YAML) or syscalls (`syscalls` in Python, XML, or YAML).
If you have installed the kernel tracer, use kernel tracing, and still encounter an error here, make sure to [add your user to the `tracing` group](#tracing).

## Tracing in snapshot mode

By default, tracing sessions write trace data continuously to disk.
Tracing sessions in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode) store trace data in memory and only write to disk when a [snapshot is taken](https://lttng.org/docs/v2.13/#doc-taking-a-snapshot).
When memory buffers fill up, the oldest data is discarded, maintaining a rolling history whose size can be controlled by configuring sub-buffer size.

> This is useful for capturing trace data only when something interesting occurs, avoiding continuous disk writes and thus lowering the runtime performance impact even more.

Reviewer: Minor suggestion, feel free to ignore: I would specifically use the words "flight recorder" here, because people are often familiar with this concept (which I assume is why LTTng's docs mention that too)

```suggestion
This "flight recorder" mode is useful for capturing trace data only when something interesting occurs, avoiding continuous disk writes and thus lowering the runtime performance impact even more.
```

Author: I'm adding this

> [!NOTE]
> In snapshot mode, high-frequency runtime events may overwrite initialization trace data in memory before a snapshot is taken, potentially making the trace data unusable.
> This is more likely to occur if sub-buffer sizes are too small.
> Support for configuring separate channels for initialization and runtime tracepoints is planned to address this limitation (see [#199](https://github.com/ros2/ros2_tracing/issues/199)).

## Dual session tracing

Dual session mode solves the problem of losing initialization trace data by using two separate tracing sessions: one for initialization events in [snapshot mode](https://lttng.org/docs/v2.13/#doc-tracing-session-mode), and another normal tracing session for runtime events.

> This allows starting to record trace data at any point without losing initialization data.

Reviewer: I would still use "actively" here to differentiate between the initialization snapshot session (which exists just in case and doesn't necessarily get recorded) and actively/intentionally starting to trace at runtime:

```suggestion
This allows starting to actively record trace data at any point without losing initialization data.
```

However, some settings need to be tuned for it to be fully real-time safe and for performance to be optimal for your use-case:

* timers[^rt-1]: use [read timer](https://lttng.org/docs/v2.13/#doc-channel-read-timer) to avoid a write(2) call

> * sub-buffer[^rt-1] count and size:

Reviewer: Why is this being removed?

Author: I've removed this because we can now configure both sub-buffer size using `Trace` action, but I realized that we still can't do this using `ros2 trace` command. I'm undoing this change.

Reviewer: ah! I guess we can update this to reflect that we can configure it through the `Trace` action but not the `ros2 trace` command, but yeah we can just revert this change for now
